doc/decisions/error_message.md

```diff
@@ -0,0 +1,215 @@
+They can simply take the current subclass 01100 and make "permission" be 01110 and "existence" be 01120. This will also allow
+backwards compatibility by applications just checking for resource errors in general.
+Splitting/merging/rearranging any category should only be done by a decision (such as this file here) because elektra developers
+should not be able to generate a new category as they wish because it would lead to the same proliferation of errors as we have now.
+
+The classification as of now is done with the following concept in mind:
+
+- Permanent errors
+    Generally errors which require changes by a user which are not getting fixed by eg. retrying.
+- Resource
+    The class of errors which require certain permissions (opening a file) or where some directory/file is missing.
+    One possible reaction to these kind of errors is to try a different directory/file or create a missing resource.
+- Parsing
+    Errors while parsing files such as ini/yml/etc.
+    Reactions could be to trim line endings/ try different encodings or even try another input at all.
+- Installation
+    Errors in the installation process such as missing plugins or more generic errors in beneath
+    the plugins/core (see #144,145,151 for example) which cannot be fixed programmatically but needs
+    further investigation. 
+- Logical
+    These errors indicate a bug in beneath elektra and should never have happened.
+- Conflict
+    Conflicts indicate temporary problems but also could be harmful such as overwriting a file after a merge conflict.
+- Timeout
+    These errors indicate temporary failures such as connection timeouts or not enough space available. Retrying at a later point in time (or
+    immediately) will most likely fix the problem. 
+- Validation
+    These errors usually indicate wrong values for certain keys such as an invalid port number or wrong types. Such errors require user action and
+    retry with a different value.
+- Syntactic
+    Syntactic failures indicate a wrong notation or missing characters for example. They require a certain form which was given by the user.
+- Semantic
+    Semantic errors are failures which indicate a different expectation of the program than from the user concerning the meaning of a value.
+    One example would be a wrong type or out-of-range errors.
+
+These categories are chosen because they can help developers to react programmatically and cover the majority of use cases to our present knowledge.
+If there is ever the need for another reaction based category, it can be extended very easily.
+
+The API for the errors will be extended as following:
+
+```
+// Already present
+ELEKTRA_SET_ERROR (nr, parentKey, message);
+ELEKTRA_SET_ERRORF(nr, parentKey, message, ...);
+ELEKTRA_ADD_WARNING(nr, parentKey, message);
+ELEKTRA_ADD_WARNINGF(nr, parentKey, message, ...)
+
+// Extended API
+ELEKTRA_SET_ERROR_WITH_SOLUTION(nr, parentKey, solution, message)
```

Comment:
With this approach we would not have format specifiers for solutions and it would also blow up the size of kdberrors.h (maybe you even find a way how to make the file shorter).

What about having a macro `ELEKTRA_ADD_SOLUTION` which adds a solution message to the last error/warning? This would also allow something like src/tools/kdb/mountbase.cpp very bottom.

Reply:
Concerning the "blow up" part I already talked about with @kodebach. With some refactoring we could drastically reduce the code duplication generated by the cpp file.

Yes, we could add another macro which could do that but would also break the atomicity then. Eg. plugin a and b throw an independent error. Plugin a also wants to add a solution. First plugin a writes to the parent key. Then plugin b follows immediately. Plugin a then adds its solution which would not fit together then

Reply:
> maybe you even find a way how to make the file shorter

I got some ideas about this while working on the highlevel code generation. I already sent them to @Piankero, but I will post it here again, in case you have any input:

**Solution 1:** 1 macro + 1 short static inline function per error code

We start with a header file that contains the basic macros (`ELEKTRA_SET_ERROR` etc):
```c
#define ELEKTRA_SET_ERROR(code, key, message, <...>) elektraSetError##code (key, message, <...>)
```

For each error code we append something like the following:
```c
#define ELEKTRA_ERROR_SOMETHING 01001

static inline void elektraSetError01001 (Key * key, const char * message, <...>) {
    keySetMeta(key, "error", "01001");
   [...]
    elektraSetErrorInformation (message, <...>);
}
```

`elektraSetErrorInformation` sets the common attributes of an error. It should be declared in the generated header (or one that is included in the generated file), and defined in a separate C file, so that we don't duplicate the code everytime `kdberrors.h` is included.

The `<...>` part is a placeholder for the remaining dynamic parameters needed for error creation. The `[...]` part is a placeholder for the remaining static properties of an error given in the specification.

**Solution 2:** 2(?) macros per error code

By abusing the preprocessor a bit, we could also do everything without the inline functions.
The reason the inline functions exist in the current (AFAIK) and the above solution, is so that we get compiler errors, if `ELEKTRA_SET_ERROR` is used with unspecified error codes. The following fulfills the same purpose:

Like above we start with a header containing the basic macros. But instead of simply calling the inline function we do something like:

```c
#define ELEKTRA_ERROR_PROPS(code) ELEKTRA_ERROR_PROPS_##code

#define ELEKTRA_SET_ERROR(code, key, message, <...>) \
    do {                                             \
        ELEKTRA_ERROR_PROPS(code);                   \
        elektraSetErrorInformation(message, <...>);  \
   } while(0)
```

```c
#define ELEKTRA_ERROR_SOMETHING 01001

#define ELEKTRA_ERROR_PROPS_01001(key)     \
        keySetMeta(key, "error", "01001"); \
        [...]
```
We might need additional helper macros so the substitution is done correctly

Beacuse `ELEKTRA_ERROR_PROPS_##code` only exists for specified we will still get a compiler error for undefined error codes. The advantage of using macros is that they are definitely inlined, although the static inline functions should also be inlined in optimized builds.

> What about having a macro ELEKTRA_ADD_SOLUTION which adds a solution message to the last error/warning?

I think that is better than having `ELEKTRA_SET_ERROR_WITH_SOLUTION`,  but I would prefer a function (should reduce binary size, if not static) instead of a macro, unless you know a reason why a function wouldn't work.

Reply:
> Eg. plugin a and b throw an independent error.

Plugins are always called sequentially by Elektra, so this situation cannot normally arise. Elektra is also explicitly not thread safe, so if parallelism occurs the developer has to handle all problems that may arise.

As long as the plugin doesn't try to do something stupid like setting an error in kdbGet and the solution in kdbSet, everything should be fine.

Reply:
> Plugins are always called sequentially by Elektra, so this situation cannot normally arise. 

There are some situations, e.g. if the plugins within the "error" position emit further errors or multiple notification fail (which happens after commits).

Of course we can prohibit to emit such errors in such situations but that should be explained in guidelines hopefully written by @Piankero

Reply:
I like @kodebach idea to reduce the size of kdberrors.h. This should also be part of the new concept. But as said below, we need a project or some other means to keep track of these proposals.